### PR TITLE
[flang] Add parsing of DO CONCURRENT REDUCE clause

### PR DESCRIPTION
**Pull request by khaki3**
Derived from #92480. This PR supports parsing of the DO CONCURRENT REDUCE clause in Fortran 2023. Following the style of the OpenMP parser in MLIR, the front end accepts both arbitrary operations and procedures for the REDUCE clause. But later Semantics can notify type errors and resolve procedure names.

### CODE REVIEW

**Review by clementval — APPROVED** [review]
LGTM! Thanks Matsu to split the PR.

**Review by klausler — CHANGES_REQUESTED** [review]
I have scanned this briefly and there are many things that need fixing here.  It's going to take an hour or se to write them up; please stand by.

File: flang/lib/Semantics/check-do-forall.cpp
Comment (klausler) [inline flang/lib/Semantics/check-do-forall.cpp:691]: `bool supportedIdentifier{true};`

File: flang/lib/Semantics/check-do-forall.cpp
Comment (klausler) [inline flang/lib/Semantics/check-do-forall.cpp:698]: `parser::Name`, we know that we're in `Fortran::`.

File: flang/lib/Semantics/check-do-forall.cpp
Comment (klausler) [inline flang/lib/Semantics/check-do-forall.cpp:700]: Better error recovery: don't assume that `x.symbol` is not a null pointer.

File: flang/lib/Semantics/check-do-forall.cpp
Comment (klausler) [inline flang/lib/Semantics/check-do-forall.cpp:708]: What does "composite type" mean in this context?

File: flang/lib/Semantics/resolve-names.cpp
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:6471]: Braced initialization, please.

File: flang/lib/Semantics/resolve-names.cpp
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:6472]: When updating a comment that is a textual reference into the standard, update any subclause, requirement, or constraint number to the latest standard, and don't bother retaining the F'2018 number: `// F'2023 C1130` is fine here.

(References to standards prior to F'2018 should be retained, they can appear in explanations of how the language has changed over time.)

File: flang/lib/Semantics/resolve-names.cpp
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:6475]: The parentheses are needless here in the predicate of the conditional expression.

File: flang/lib/Semantics/resolve-names.cpp
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:6893]: It would be less confusing if your updated `DeclareLocalEntity()` set the flag now, if that is possible.

File: flang/test/Semantics/resolve124.f90
Comment (klausler) [inline flang/test/Semantics/resolve124.f90:23]: Naming the unacceptable type and the "suitable" type(s) would be more useful to the programmer.

**Review by klausler — CHANGES_REQUESTED** [review]

File: flang/include/flang/Parser/parse-tree.h
Comment (khaki3) [reply flang/include/flang/Parser/parse-tree.h:2251]: I'll repeat a comment sent personally to you before.

>I tried an implementation in a similar way to AccReductionOperator::Operator. It cannot tell the reason for parsing failures. The name of AccReductionOperator::Operator is redundant. I tried to add ReduceOperation::Add directly (without ::Operator::), but the OpenACC version has a sourced parser. Defining 'source' in 'ReduceOperation' does not match with other code.

There was no CUF REDUCE support that time. Should I check it?

File: flang/lib/Parser/executable-parsers.cpp
Comment (khaki3) [reply flang/lib/Parser/executable-parsers.cpp:270]: "REDUCE (" allows any whitespace?

File: flang/lib/Parser/unparse.cpp
Comment (khaki3) [reply flang/lib/Parser/unparse.cpp:1043]: Nice!

File: flang/lib/Semantics/check-do-forall.cpp
Comment (khaki3) [reply flang/lib/Semantics/check-do-forall.cpp:708]: I believe reduction variables can have derived types. The same logic is found here: https://github.com/llvm/llvm-project/blob/main/flang/lib/Semantics/check-acc-structure.cpp#L700

I will change it from `composite type` to `derived type`.

File: flang/test/Semantics/resolve124.f90
Comment (khaki3) [reply flang/test/Semantics/resolve124.f90:23]: Sure, I will put the list of suitable types.

File: flang/include/flang/Parser/parse-tree.h
Comment (klausler) [reply flang/include/flang/Parser/parse-tree.h:2251]: Yes.  And you can handle the error recovery of an enumerated operator in the parser rather than allowing the parse tree to hold an overly general procedure designator.

File: flang/lib/Parser/executable-parsers.cpp
Comment (klausler) [reply flang/lib/Parser/executable-parsers.cpp:270]: It allows a space between the keyword and the parenthesis.

File: flang/lib/Semantics/check-do-forall.cpp
Comment (klausler) [reply flang/lib/Semantics/check-do-forall.cpp:708]: See C1132 in F'2023.

File: flang/lib/Semantics/check-do-forall.cpp
Comment (khaki3) [reply flang/lib/Semantics/check-do-forall.cpp:686]: It does not allow `static` because of the use of `context_`.

**Comment by khaki3** [comment]
I appreciate your comments. The current version is sharing ReductionOperator among FIR/ACC/CUF. The test was also refreshed.

File: flang/lib/Semantics/check-do-forall.cpp
Comment (klausler) [inline flang/lib/Semantics/check-do-forall.cpp:691]: `supportedIdentifier` would be consistent.

File: flang/lib/Semantics/check-do-forall.cpp
Comment (klausler) [inline flang/lib/Semantics/check-do-forall.cpp:694]: `TypeMismatch`

File: flang/lib/Semantics/check-do-forall.cpp
Comment (klausler) [inline flang/lib/Semantics/check-do-forall.cpp:696]: Please leave error messages as long contiguous character literals, even if they exceed the 80 character limit.  Otherwise they would be much harder to find with grep.

File: flang/lib/Semantics/resolve-names.cpp
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:2266]: You can probably omit `Message{` and its corresponding `}` here.

File: flang/lib/Semantics/resolve-names.cpp
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:6490]: Please don't break up the messages here and following.

**Review by klausler — CHANGES_REQUESTED** [review]

File: flang/lib/Parser/openacc-parsers.cpp
Comment (clementval) [inline flang/lib/Parser/openacc-parsers.cpp:95]: Maybe update the comment here to reflect OpenACC/Do concurrent sections number

File: flang/include/flang/Parser/parse-tree.h
Comment (khaki3) [reply flang/include/flang/Parser/parse-tree.h:2246]: The OpenMP design stems from ClauseT.h in the LLVM frontend. It allows a unique reduction identifier, so we should keep the current parser.
https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Frontend/OpenMP/ClauseT.h#L258

**Comment by khaki3** [comment]
Thank you for the reviews. I dealt with all the points. Please check my update again.

**Comment by clementval** [comment]
> Thank you for the reviews. I dealt with all the points. Please check my update again.

OpenACC changes look good to me.

File: flang/lib/Semantics/resolve-names.cpp
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:6474]: Declare `const char *specName{... ? "LOCAL_INIT" : "LOCAL"};` after line 6470 and then use `specName` as the argument, it will be easier to understand than this repetitive conditional suffix.

File: flang/lib/Semantics/resolve-names.cpp
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:6903]: `const auto &name`

**Review by klausler — APPROVED** [review]

**Comment by clementval** [comment]
Still LGTM. Do you have commit access or do you need someone to merge your change?

**Comment by khaki3** [comment]
It seems that I need another approval:
>This workflow requires approval from a maintainer.

**Comment by clementval** [comment]
> It seems that I need another approval:
> 
> > This workflow requires approval from a maintainer.

First time that I see this button. I just approve and run.

**Pull request merged**
